src/payment/strategies/zip/zip-payment-strategy.ts

```diff
@@ -0,0 +1,119 @@
+
+import { Zip, ZipModalEvent } from './zip';
+import ZipScriptLoader from './zip-script-loader';
+
+export default class ZipPaymentStrategy implements PaymentStrategy {
+    private _methodId?: string;
+    private _paymentMethod?: PaymentMethod;
+    private _zipClient?: Zip;
+
+    constructor(
+        private _store: CheckoutStore,
+        private _orderActionCreator: OrderActionCreator,
+        private _paymentActionCreator: PaymentActionCreator,
+        private _paymentMethodActionCreator: PaymentMethodActionCreator,
+        private _zipScriptLoader: ZipScriptLoader
+    ) { }
+
+    initialize(options: PaymentInitializeOptions): Promise<InternalCheckoutSelectors> {
+        this._methodId = options.methodId;
+
+        if (!this._methodId) {
+            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+        }
+
+        return this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(this._methodId))
+            .then(state => {
+                if (!this._methodId) {
+                    throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+                }
+
+                this._paymentMethod = state.paymentMethods.getPaymentMethod(this._methodId);
+
+                if (!this._paymentMethod || !this._paymentMethod.clientToken) {
+                    throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+                }
+
+                return this._zipScriptLoader.load()
+                    .then(zip => {
+                        this._zipClient = zip;
+
+                        return this._store.getState();
+                    });
+            });
+    }
+
+    deinitialize(options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        this._paymentMethod = undefined;
+        this._zipClient = undefined;
+
+        return Promise.resolve(this._store.getState());
+    }
+
+    execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        const order = { useStoreCredit: payload.useStoreCredit };
+
+        return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
+            .then(() => new Promise((resolve, reject) => {
+                if (!this._zipClient) {
+                    throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+                }
+
+                this._zipClient.Checkout.init({
+                    onComplete: ({ checkoutId, state }) => {
+                        if (state === ZipModalEvent.CancelCheckout) {
+                            return reject(new PaymentMethodCancelledError());
+                        }
+
+                        if (state === ZipModalEvent.CheckoutApproved) {
+                            if (!this._methodId) {
+                                return reject(new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized));
```

Comment:
I would probably just get the `methodId` from `payload` instead of storing it as an instance variable. So the whole thing could be written like this

```ts
const { payment, ...order } = payload;
const { _zipClient: zipClient } = this;

if (!payment) {
    throw new PaymentArgumentInvalidError();
}

// NOTE: I've moved the check here because at this point we can already determine if payment method is initialised or not.
if (!zipClient) {
    throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
}

return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
    .then(() => new Promise((resolve, reject) => {
        zipClient.Checkout.init({
            onComplete: ({ checkoutId, state }) => {
                if (state === ZipModalEvent.CheckoutApproved && checkoutId) {
                    return resolve(checkoutId);
                }

                if (state === ZipModalEvent.CancelCheckout) {
                    return reject(new PaymentMethodCancelledError());
                }

                return reject(new PaymentMethodInvalidError());
            },

            onCheckout: () => {
                // Pass the clientToken to ZipPay here
            },
        });
    })
    // NOTE: I've moved the call here to reduce code nesting.
    .then(nonce => this._store.dispatch(this._paymentActionCreator.submitPayment({
        ...payment,
        paymentData: { nonce },
    }));
```

Reply:
I copied the zipClient null check before `submitOrder`, to it can't be submitted  if the client is not loaded.
But it seems that the arrow function of  the `dispatch`causes the linter to ignore the null check from before (see image) , so I had to duplicate the code as a workaround, should I leave that change or is there something else I can do?
![Screen Shot 2019-03-25 at 10 26 27 AM](https://user-images.githubusercontent.com/35502707/54942670-3387c280-4ef5-11e9-8b9e-0fdafc536882.png)

Reply:
Have you tried what I suggest above, which is to hold a direct reference to the client instance and use it within the method?